File: web-sdk/src/data/services/yalo-message/yalo-message-service-websocket.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import { Err, Ok, type Result } from '@domain/common/result';
import {
  ConnectionAck,
  ConnectionAckType,
  PollMessageItem,
  SdkMessage,
} from '@domain/models/events/external_channel/in_app/sdk/sdk_message';
import type { TokenRepository } from '@data/repositories/token/token-repository';
import type {
  MessageCallback,
  YaloMessageService,
} from './yalo-message-service';

const INITIAL_BACKOFF_MS = 1000;
const MAX_BACKOFF_MS = 30000;
const ACK_TIMEOUT_MS = 10000;

export class YaloMessageServiceWebSocket implements YaloMessageService {
  private readonly _wsUrl: string;
  private readonly _tokenRepository: TokenRepository;
  private _socket?: WebSocket;
  private _callback?: MessageCallback;
  private _reconnectAttempt = 0;
  private _reconnectTimeout?: ReturnType<typeof setTimeout>;
  private _ackTimeout?: ReturnType<typeof setTimeout>;
  private _running = false;
  private _pendingFrames: string[] = [];
  private _connectionAcked = false;

  constructor(baseUrl: string, tokenRepository: TokenRepository) {
    this._wsUrl = `wss://${baseUrl}/websocket/v1/connect/webchat`;
    this._tokenRepository = tokenRepository;
  }

  subscribe(callback: MessageCallback): void {
    this._callback = callback;
    this._running = true;
    this._connect();
  }

  unsubscribe(): void {
    this._running = false;
    this._callback = undefined;
    this._reconnectAttempt = 0;
    this._pendingFrames = [];
    this._connectionAcked = false;
    if (this._reconnectTimeout) {
      clearTimeout(this._reconnectTimeout);
      this._reconnectTimeout = undefined;
    }
    this._clearAckTimeout();
    if (this._socket) {
      this._socket.close();
      this._socket = undefined;
    }
  }

  async sendMessage(message: SdkMessage): Promise<Result<void>> {
    if (!this._running) {
      return new Err(new Error('WebSocket is not connected'));
    }
    let frame: string;
    try {
      frame = JSON.stringify(SdkMessage.toJSON(message));
    } catch (e) {
      return new Err(e instanceof Error ? e : new Error(String(e)));
    }
    if (
      !this._connectionAcked ||
      this._socket?.readyState !== WebSocket.OPEN
    ) {
      this._pendingFrames.push(frame);
      return new Ok(undefined);
    }
    try {
      this._socket.send(frame);
      return new Ok(undefined);
    } catch (e) {
      return new Err(e instanceof Error ? e : new Error(String(e)));
    }
  }

  private _clearAckTimeout(): void {
    if (this._ackTimeout) {
      clearTimeout(this._ackTimeout);
      this._ackTimeout = undefined;
    }
  }

  private _flushPending(): void {
    const pending = this._pendingFrames;
    this._pendingFrames = [];
    for (const frame of pending) {
      try {
        this._socket?.send(frame);
      } catch {
        // The caller already received Ok at enqueue time.
      }
    }
  }

  private async _connect(): Promise<void> {
    if (!this._running) return;
    if (this._socket) return;

    const tokenResult = await this._tokenRepository.getToken();
    if (!tokenResult.ok) {
      this._scheduleReconnect();
      return;
    }

    const url = `${this._wsUrl}?token=${encodeURIComponent(tokenResult.value)}`;
    const socket = new WebSocket(url);
    this._socket = socket;

    socket.addEventListener('open', () => {
      this._reconnectAttempt = 0;
      this._ackTimeout = setTimeout(() => {
        this._ackTimeout = undefined;
        socket.close();
      }, ACK_TIMEOUT_MS);
    });

    socket.addEventListener('message', (event: MessageEvent) => {
      if (typeof event.data !== 'string') {
        return;
      }
      let parsed: unknown;
      try {
        parsed = JSON.parse(event.data);
      } catch {
        return;
      }
      if (typeof parsed !== 'object' || parsed === null) {
        return;
      }
      try {
        if (!this._connectionAcked) {
          const ack = ConnectionAck.fromJSON(parsed);
          if (ack.type === ConnectionAckType.CONNECTION_ACK_TYPE_CONNECTION_ACK) {
            this._connectionAcked = true;
            this._clearAckTimeout();
            this._flushPending();
          }
          return;
        }
        const item = PollMessageItem.fromJSON(parsed);
        this._callback?.(item);
      } catch {
        // ignore malformed frames
      }
    });

    socket.addEventListener('error', () => {
      socket.close();
    });

    socket.addEventListener('close', () => {
      if (this._socket === socket) {
        this._socket = undefined;
      }
      this._connectionAcked = false;
      this._clearAckTimeout();
      if (this._running) {
        this._scheduleReconnect();
      }
    });
  }

  private _scheduleReconnect(): void {
    if (!this._running || this._reconnectTimeout) return;
    const delay = Math.min(
      MAX_BACKOFF_MS,
      INITIAL_BACKOFF_MS * 2 ** this._reconnectAttempt
    );
    this._reconnectAttempt++;
    this._reconnectTimeout = setTimeout(() => {
      this._reconnectTimeout = undefined;
      this._connect();
    }, delay);
  }
}

File: flutter-sdk/lib/src/data/services/yalo_message/yalo_message_service_websocket.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:async';
import 'dart:convert';

import 'package:logging/logging.dart';
import 'package:web_socket_channel/io.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:yalo_chat_flutter_sdk/src/common/result.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message/yalo_message_service.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/token_entry.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/yalo_message_auth_service.dart';
import 'package:yalo_chat_flutter_sdk/src/domain/models/events/external_channel/in_app/sdk/sdk_message.pb.dart';

typedef WebSocketChannelFactory = WebSocketChannel Function(Uri uri);

class YaloMessageServiceWebSocket implements YaloMessageService {
  static const Duration _initialBackoff = Duration(seconds: 1);
  static const Duration _maxBackoff = Duration(seconds: 30);
  static const Duration _ackTimeout = Duration(seconds: 10);

  final String _wsUrl;
  final YaloMessageAuthService _authService;
  final WebSocketChannelFactory _channelFactory;
  final Logger log = Logger('YaloMessageServiceWebSocket');

  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _subscription;
  StreamController<PollMessageItem>? _controller;
  Timer? _reconnectTimer;
  Timer? _ackTimer;
  int _reconnectAttempt = 0;
  bool _running = false;
  bool _paused = false;
  bool _connectionAcked = false;
  final List<String> _pendingFrames = [];

  YaloMessageServiceWebSocket({
    required String baseUrl,
    required YaloMessageAuthService authService,
    WebSocketChannelFactory? channelFactory,
  }) : _wsUrl = 'wss://$baseUrl/websocket/v1/connect/inapp',
       _authService = authService,
       _channelFactory =
           channelFactory ?? ((uri) => IOWebSocketChannel.connect(uri));

  @override
  Stream<PollMessageItem> messages() {
    final controller = _controller ??=
        StreamController<PollMessageItem>.broadcast();
    if (!_running) {
      _running = true;
      _connect();
    }
    return controller.stream;
  }

  @override
  void pause() {
    if (!_running) {
      return;
    }
    log.info('Connection paused');
    _running = false;
    _paused = true;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _clearAckTimer();
    _subscription?.cancel();
    _subscription = null;
    _channel?.sink.close();
    _channel = null;
    _connectionAcked = false;
  }

  @override
  void resume() {
    if (!_paused || _controller == null) {
      return;
    }
    log.info('Connection resumed');
    _paused = false;
    _running = true;
    _reconnectAttempt = 0;
    _connect();
  }

  @override
  Future<Result<Unit>> sendSdkMessage(SdkMessage message) async {
    if (!_running) {
      return Result.error(Exception('WebSocket is not connected'));
    }
    final String frame;
    try {
      frame = jsonEncode(message.toProto3Json());
    } on Exception catch (e) {
      return Result.error(e);
    }
    final channel = _channel;
    if (!_connectionAcked || channel == null) {
      _pendingFrames.add(frame);
      return Result.ok(Unit());
    }
    try {
      channel.sink.add(frame);
      return Result.ok(Unit());
    } on Exception catch (e) {
      return Result.error(e);
    }
  }

  @override
  void dispose() {
    _running = false;
    _paused = false;
    _reconnectAttempt = 0;
    _connectionAcked = false;
    _pendingFrames.clear();
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _clearAckTimer();
    _subscription?.cancel();
    _subscription = null;
    _channel?.sink.close();
    _channel = null;
    _controller?.close();
    _controller = null;
  }

  void _clearAckTimer() {
    _ackTimer?.cancel();
    _ackTimer = null;
  }

  void _flushPending() {
    final channel = _channel;
    if (channel == null) {
      return;
    }
    final pending = List<String>.from(_pendingFrames);
    _pendingFrames.clear();
    for (final frame in pending) {
      try {
        channel.sink.add(frame);
      } on Exception {
        // The caller already received Ok at enqueue time.
      }
    }
  }

  Future<void> _connect() async {
    if (!_running || _channel != null) {
      return;
    }

    final authResult = await _authService.auth();
    if (authResult case Error()) {
      log.warning('Auth failed, scheduling reconnect');
      _scheduleReconnect();
      return;
    }
    if (!_running) {
      return;
    }
    final entry = (authResult as Ok<TokenEntry>).result;

    final WebSocketChannel channel;
    try {
      final uri = Uri.parse(
        '$_wsUrl?token=${Uri.encodeComponent(entry.accessToken)}',
      );
      channel = _channelFactory(uri);
    } on Exception catch (e) {
      log.warning('Failed to open websocket', e);
      _scheduleReconnect();
      return;
    }

    _channel = channel;
    _reconnectAttempt = 0;
    _ackTimer = Timer(_ackTimeout, () {
      _ackTimer = null;
      channel.sink.close();
    });

    _subscription = channel.stream.listen(
      _onFrame,
      onDone: _onDone,
      onError: _onError,
      cancelOnError: true,
    );
  }

  void _onFrame(dynamic data) {
    if (data is! String) {
      return;
    }
    final Map<String, dynamic> json;
    try {
      final decoded = jsonDecode(data);
      if (decoded is! Map<String, dynamic>) {
        return;
      }
      json = decoded;
    } on FormatException {
      return;
    }
    try {
      if (!_connectionAcked) {
        final ack = ConnectionAck.create()
          ..mergeFromProto3Json(json, ignoreUnknownFields: true);
        if (ack.type == ConnectionAckType.CONNECTION_ACK_TYPE_CONNECTION_ACK) {
          _connectionAcked = true;
          _clearAckTimer();
          _flushPending();
        }
        return;
      }
      final item = PollMessageItem.create()..mergeFromProto3Json(json);
      _controller?.add(item);
    } on Exception catch (e) {
      log.fine('Ignoring malformed websocket frame', e);
    }
  }

  void _onError(Object error) {
    log.warning('WebSocket error', error);
    _handleClose();
  }

  void _onDone() {
    _handleClose();
  }

  void _handleClose() {
    _subscription?.cancel();
    _subscription = null;
    _channel = null;
    _connectionAcked = false;
    _clearAckTimer();
    if (_running) {
      _scheduleReconnect();
    }
  }

  void _scheduleReconnect() {
    if (!_running || _reconnectTimer != null) {
      return;
    }
    final ms = (_initialBackoff.inMilliseconds * (1 << _reconnectAttempt))
        .clamp(_initialBackoff.inMilliseconds, _maxBackoff.inMilliseconds);
    _reconnectAttempt++;
    _reconnectTimer = Timer(Duration(milliseconds: ms), () {
      _reconnectTimer = null;
      _connect();
    });
  }
}
